File: semantic-model/SemForge-SDK/vscode/src/project.js
/*
 * The project itself: which package this is, how it is set up, what it holds.
 *
 * The other three views answer what the package SAYS -- its constraints, its
 * data, its vocabulary. None answered what it IS. A window showed a name and
 * nothing else, and two directories called `test` are not the same project:
 * there were no settings to read, no status to check, and no way to change
 * either without finding semforge.yaml by hand.
 *
 * So this view sits above the other three and holds the answers, with each
 * setting editable in place. Editing writes one line of semforge.yaml and
 * leaves every comment in it -- those comments are the documentation.
 */

const vscode = require('vscode');

const { noPackageMessage } = require('./locate');
const { showLocation } = require('./reveal');

class ProjectTreeNode {
  constructor(key, raw, packageUri) {
    this.key = key;
    this.raw = raw;
    this.packageUri = packageUri;
  }
}

function keyOf(raw, parentKey, position) {
  return `${parentKey}/${position}:${raw.kind}:${raw.key || raw.label}`;
}

class ProjectTreeProvider {
  constructor(clientHolder) {
    this.clientHolder = clientHolder;
    this._onDidChangeTreeData = new vscode.EventEmitter();
    this.onDidChangeTreeData = this._onDidChangeTreeData.event;
    this.nodes = new Map();
    this.parents = new Map();
  }

  refresh(uri) {
    if (uri) {
      this.uri = uri;
    }
    this._onDidChangeTreeData.fire();
  }

  wrap(raw, key) {
    const existing = this.nodes.get(key);
    if (existing) {
      existing.raw = raw;
      existing.packageUri = this.uri;
      return existing;
    }
    const made = new ProjectTreeNode(key, raw, this.uri);
    this.nodes.set(key, made);
    return made;
  }

  getParent(node) {
    return this.parents.get(node.key);
  }

  getTreeItem(node) {
    const raw = node.raw;
    const children = raw.children || [];
    const item = new vscode.TreeItem(
      raw.label || raw.kind,
      children.length
        ? raw.kind === 'group' || raw.kind === 'project'
          ? vscode.TreeItemCollapsibleState.Expanded
          : vscode.TreeItemCollapsibleState.Collapsed
        : vscode.TreeItemCollapsibleState.None
    );
    // The value is the point of the row, so it goes where the eye lands; what
    // the setting DECIDES is the tooltip, because it is a paragraph.
    item.description = raw.value || '';
    // `editable` decides, for every kind. Keying this on the kind is what left
    // the pencil off rows that carry a value -- twice.
    item.contextValue = raw.editable ? 'setting' : raw.kind;

    if (raw.kind === 'group' || raw.kind === 'project') {
      item.iconPath = new vscode.ThemeIcon('project');
    } else if (raw.severity) {
      item.iconPath = new vscode.ThemeIcon('warning');
    } else if (raw.kind === 'namespaces' || raw.kind === 'namespaceEntry') {
      item.iconPath = new vscode.ThemeIcon('symbol-namespace');
    } else if (raw.editable) {
      item.iconPath = new vscode.ThemeIcon('settings-gear');
    } else {
      item.iconPath = new vscode.ThemeIcon('info');
    }

    const lines = [];
    if (raw.detail) {
      lines.push(raw.detail);
    }
    if (raw.doc && raw.doc !== raw.detail) {
      lines.push(raw.doc);
    }
    if (raw.definedAt) {
      lines.push(raw.definedAt);
    }
    if (lines.length) {
      item.tooltip = lines.join('\n\n');
    }
    return item;
  }

  async getChildren(node) {
    if (node) {
      return (node.raw.children || []).map((child, position) => {
        const key = keyOf(child, node.key, position);
        this.parents.set(key, node);
        return this.wrap(child, key);
      });
    }
    const client = this.clientHolder.client;
    if (!client) {
      this.view.message = 'The language server is not running — run ' +
        '"SemForge: Doctor".';
      return [];
    }
    if (!this.uri) {
      this.view.message = noPackageMessage();
      return [];
    }
    let result;
    try {
      result = await client.sendRequest('semforge/project', { uri: this.uri });
    } catch (error) {
      this.view.message = `SemForge: ${error.message || error}`;
      return [];
    }
    if (result.error) {
      this.view.message = `SemForge: ${result.error}`;
      return [];
    }
    this.view.message = undefined;
    this.parents = new Map();
    return (result.roots || []).map((raw, position) =>
      this.wrap(raw, keyOf(raw, '', position)));
  }
}

function register(context, clientHolder, session, onChanged) {
  const provider = new ProjectTreeProvider(clientHolder);
  const view = vscode.window.createTreeView('semforgeProject', {
    treeDataProvider: provider
  });
  provider.view = view;
  context.subscriptions.push(view);

  // Selecting a row that names a line opens it. A setting's row points at its
  // own line in semforge.yaml, so reading the comment around it is one click.
  context.subscriptions.push(
    view.onDidChangeSelection(async (event) => {
      const selected = event.selection && event.selection[0];
      if (selected && selected.raw.definedAt) {
        await showLocation(selected.raw.definedAt, false);
      }
    })
  );

  provider.refresh(session.uri);
  context.subscriptions.push(
    session.onDidChange((uri) => provider.refresh(uri)),
    vscode.workspace.onDidSaveTextDocument(() => provider.refresh())
  );

  context.subscriptions.push(
    vscode.commands.registerCommand('semforge.editSetting', async (node) => {
      const raw = node && node.raw;
      if (!raw || !raw.editable) {
        return;
      }
      const value = await vscode.window.showInputBox({
        title: `${raw.label}`,
        prompt: raw.doc || raw.detail,
        value: raw.value === '—' ? '' : raw.value,
        ignoreFocusOut: true
      });
      if (value === undefined) {
        return;                       // cancelled; nothing is written
      }
      const result = await clientHolder.client.sendRequest(
        'semforge/setSetting',
        { uri: node.packageUri, key: raw.key, value }
      );
      if (!result || !result.ok) {
        vscode.window.showErrorMessage(
          `SemForge: ${(result && result.error) || 'the setting was not written'}`
        );
        return;
      }
      provider.refresh();
      if (onChanged) {
        // The context and the entity root change what the other views mean,
        // so they are re-read rather than left showing the old answer.
        onChanged();
      }
      await showLocation(`${result.file}:${result.line}`, false);
    }),

    vscode.commands.registerCommand('semforge.addNamespace', async (node) => {
      // Prefixes are a package-wide table, agreed once. There was no way to
      // add to it from the editor at all -- you found semforge.yaml and typed.
      const prefix = await vscode.window.showInputBox({
        title: 'New namespace prefix',
        prompt: 'The name this package uses for it, e.g. plant',
        validateInput: (text) =>
          /^[A-Za-z_][\w.-]*$/.test((text || '').replace(/:$/, ''))
            ? undefined
            : 'A letter or underscore, then letters, digits, dots, ' +
              'underscores or hyphens.'
      });
      if (!prefix) {
        return;
      }
      const namespace = await vscode.window.showInputBox({
        title: `What does ${prefix}: mean?`,
        prompt: 'The namespace IRI. It has to end in "/", "#" or ":", or a ' +
          'term appended to it runs into the last segment.',
        value: 'https://'
      });
      if (!namespace) {
        return;
      }
      const made = await clientHolder.client.sendRequest(
        'semforge/addNamespace',
        { uri: (node && node.packageUri) || provider.uri, prefix, namespace }
      );
      if (!made || !made.ok) {
        vscode.window.showErrorMessage(
          `SemForge: ${(made && made.error) || 'the prefix was not defined'}`
        );
        return;
      }
      provider.refresh();
      if (onChanged) {
        onChanged();
      }
      await showLocation(`${made.file}:${made.line}`, false);
    }),

    vscode.commands.registerCommand('semforge.removeNamespace', async (node) => {
      const raw = node && node.raw;
      if (!raw || !raw.label) {
        return;
      }
      const uri = node.packageUri || provider.uri;
      // The server decides whether it CAN go; it also answers "it is in use"
      // as a question rather than doing it. A line that changes nothing is
      // still a line somebody wrote on purpose, and what the person clicking
      // is thinking about is that three files bind it.
      const ask = async (force) =>
        clientHolder.client.sendRequest('semforge/removeNamespace',
                                        { uri, prefix: raw.label, force });

      let gone = await ask(false);
      if (gone && gone.confirm) {
        const answer = await vscode.window.showWarningMessage(
          `Remove ${raw.label}: ?`,
          { modal: true, detail: gone.detail },
          'Remove'
        );
        if (answer !== 'Remove') {
          return;
        }
        gone = await ask(true);
      }
      if (!gone || !gone.ok) {
        vscode.window.showWarningMessage(
          `SemForge: ${(gone && gone.error) || 'the prefix was not removed'}`
        );
        return;
      }
      provider.refresh();
      if (onChanged) {
        onChanged();
      }
      vscode.window.setStatusBarMessage(
        gone.survivesVia
          ? `SemForge: ${raw.label}: removed — ${gone.survivesVia} still names it`
          : `SemForge: ${raw.label}: removed`,
        5000
      );
    }),

    vscode.commands.registerCommand('semforge.refreshProject', () =>
      provider.refresh()
    )
  );

  return provider;
}

module.exports = { register, ProjectTreeProvider };

File: semantic-model/SemForge-SDK/semforge/cooked/project.py
"""The project itself, as a tree: what this package is and how it is set up.

The three views answer what the package SAYS -- its constraints, its data, its
vocabulary. None of them answered what the package IS. Two directories called
`test` are not the same project, and nothing on screen separated them: no
settings, no status, no way to change either.

So a fourth view, above the other three, holding the answers in one place:

  * which package, spelled out -- the name it declares, the path it is at, and
    why that directory counted as one;
  * its settings, each one editable in place, keeping the comments that say
    what it decides;
  * what it is made of -- which file or directory holds each role, how many
    documents, how many shapes, entities and cases.

Every row that names a place carries `file:line`, so clicking it opens the
line rather than the file.
"""

import os
from dataclasses import dataclass, field

from rdflib.namespace import RDF, SH

from ..expect.store import examples_root, load_expectations
from ..package import config, discover


@dataclass
class ProjectNode:
    # kind: group | setting | fact | entry
    kind: str
    label: str
    detail: str = ''
    value: str = ''
    key: str = ''              # the setting this row writes, when it writes one
    doc: str = ''              # what it decides
    editable: bool = False
    defined_at: str = ''       # file:line
    severity: str = ''
    children: list = field(default_factory=list)


def _at(path, line):
    return f'{path}:{line}' if path and line else ''


def _relative(package, path):
    try:
        return os.path.relpath(path, package.path)
    except ValueError:                             # different drive on Windows
        return path


def build_project(package):
    """The project card: identity, settings, contents."""
    return [_identity(package), _settings(package), _contents(package)]


def _identity(package):
    root = os.path.abspath(package.path)
    found, why, _ = discover.describe(root)
    manifest = config.path_for(root)
    declared = config.read(root)

    name = declared.get('name') or os.path.basename(root)
    rows = [
        ProjectNode(kind='setting', label='name', value=str(name),
                    key='name', editable=True,
                    doc=next(doc for key, _, doc in config.SPEC
                             if key == 'name'),
                    detail='' if declared.get('name')
                           else 'not declared — the directory name',
                    defined_at=_at(manifest, config.locate(root, 'name'))),
        ProjectNode(kind='fact', label='path', value=root,
                    detail='where it is on disk'),
        ProjectNode(kind='fact', label='recognised by', value=why or '—',
                    detail='what made this directory a package'),
        ProjectNode(
            kind='fact', label='manifest',
            value=config.CONFIG if os.path.isfile(manifest) else '(none)',
            detail='' if os.path.isfile(manifest)
                   else 'this package has no semforge.yaml — editing a '
                        'setting creates one',
            defined_at=_at(manifest, 1) if os.path.isfile(manifest) else ''),
    ]
    node = ProjectNode(kind='group', label='Project', detail=str(name))
    node.children = rows
    return node


def _settings(package):
    root = os.path.abspath(package.path)
    manifest = config.path_for(root)
    rows = []
    for setting in config.settings(root):
        if setting.key == 'name':
            continue                      # shown under Project, where it reads
        detail = setting.doc
        severity = ''
        if not setting.value:
            # Only where absence actually costs something. A missing published
            # context is not a style question: export cannot run without it.
            if setting.key == 'context.published':
                severity = 'warning'
                detail = 'not declared — `semforge export` has nowhere to ' \
                         'point the model. ' + setting.doc
            elif setting.default:
                detail = f'not declared — {setting.default} applies. {setting.doc}'
            else:
                detail = 'not declared. ' + setting.doc
        rows.append(ProjectNode(
            kind='setting', label=setting.label,
            value=setting.value or '—', key=setting.key,
            doc=setting.doc, editable=setting.editable, detail=detail,
            severity=severity,
            defined_at=_at(manifest, setting.line)))

    # Prefixes are a package-wide table, so what the artifacts say about it
    # belongs on the row that shows it -- not in a command nobody runs.
    trouble = _prefix_findings(package)

    for collection in config.collections(root):
        detail = collection.doc
        severity = ''
        if collection.key == 'namespaces' and trouble:
            severity = 'error' if any(f.severity == 'error' for f in trouble) \
                else 'warning'
            detail = f'{len(trouble)} disagreement(s) — ' + trouble[0].message
        group = ProjectNode(
            kind='namespaces' if collection.key == 'namespaces' else 'fact',
            label=collection.label,
            value=f'{len(collection.entries)}', doc=collection.doc,
            detail=detail, severity=severity,
            defined_at=_at(manifest, config.locate(root, collection.key)))
        group.children = [
            ProjectNode(kind='entry', label=name, value=value,
                        defined_at=_at(manifest, line))
            for name, value, line in collection.entries]
        rows.append(group)

    node = ProjectNode(kind='group', label='Settings',
                       detail=f'{config.CONFIG} — click the pencil to change one')
    node.children = rows
    return node


def _prefix_findings(package):
    """What the artifacts say about the package's namespace table."""
    from ..package.prefixes import check

    try:
        return check(package)
    except Exception:                              # noqa: BLE001
        return []                                  # `semforge prefixes` tells it better


def _contents(package):
    """What the package is made of, counted."""
    shapes = len(set(package.shapes.subjects(RDF.type, SH.NodeShape)))
    properties = len(list(package.shapes.triples((None, SH.property, None))))
    classes = len(set(package.knowledge.subjects(RDF.type, None)))
    entities = len(set(package.model.subjects(RDF.type, None)))

    rows = []
    for role, plural in (('knowledge', 'document'), ('shapes', 'document'),
                         ('model', 'document')):
        files = package.files(role)
        where = _relative(package, files[0]) if len(files) == 1 else \
            os.path.dirname(_relative(package, files[0])) + os.sep
        rows.append(ProjectNode(
            kind='fact', label=role, value=where,
            detail=f'{len(files)} {plural}{"" if len(files) == 1 else "s"}',
            defined_at=f'{files[0]}:1'))

    rows.append(ProjectNode(kind='fact', label='shapes declared',
                            value=str(shapes),
                            detail=f'{properties} property constraint(s)'))
    rows.append(ProjectNode(kind='fact', label='knowledge terms',
                            value=str(classes)))
    rows.append(ProjectNode(kind='fact', label='entities in the model',
                            value=str(entities)))

    # The suite's own count, not a directory listing: examples/ also holds
    # the shared subobjects, which are not cases.
    try:
        declared = load_expectations(package.path).examples
    except Exception:                              # noqa: BLE001
        declared = []                              # the Model view reports why
    suites = sorted({example.suite for example in declared if example.suite})
    where = examples_root(package.path)
    rows.append(ProjectNode(
        kind='fact', label='test cases', value=str(len(declared)),
        detail=f'{len(suites)} suite(s): {", ".join(suites)}' if suites
               else 'none — nothing proves a constraint can fire',
        severity='' if declared else 'warning',
        defined_at=f'{where}:1' if where and os.path.isdir(where) else ''))

    node = ProjectNode(kind='group', label='Contents',
                       detail=f'{shapes} shape(s) · {entities} entity(s) · '
                              f'{len(declared)} case(s)')
    node.children = rows
    return node
